### PR TITLE
Handle asynchronous commands when PAF-lib is loaded asynchronously

### PR DESCRIPTION
**Pull request by RomainLofaso**
# Do not merge
It is only for review because it relies on another not-merged branch.

### CODE REVIEW

**Comment by bwschmidt** [comment]
if you add the glue code into the PAF lib as i commented in https://github.com/prebid/paf-mvp-implementation/pull/168 i think you can avoid the need for a queue in paf-lib

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { TestWatcher } from 'jest';
+import { isExportDeclaration } from 'typescript';
+import { Command, CommandQueue, processCommands } from '../../src/utils/queue';
+
+class CommandMocker {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:5]: Use `jest.fn()` instead (more standard)

File: paf-mvp-frontend/src/main.ts [hunk 2]
```diff
@@ -19,4 +24,15 @@ const showNotification = (type: NotificationEnum) => notificationService.showNot
 
 // TODO: avoid global declaration
 window.PAFUI ??= { promptConsent, showNotification };
-window.PAF ??= { getNewId, signPreferences, getIdsAndPreferences, refreshIdsAndPreferences, updateIdsAndPreferences };
+window.PAF ??= {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/main.ts:27]: Should be updated to cope with the situation where the object has already been initialized in the parent HTML page:
```typescript
window.PAF = {
	// Inject the "official" queue
  queue,
	// If the queue has been defined outside of this script, will be replaced by it
  ...(window.PAF ?? {})
	// The rest has to be the official methods, should not be overridden from the outside
  getNewId,
  signPreferences,
  getIdsAndPreferences,
  refreshIdsAndPreferences,
  updateIdsAndPreferences,
  generateSeed,
  registerTransmissionResponse,
  getAuditLogByTransaction,
  getAuditLogByDivId,

};
```

=> this will ensure that the object is always populated by the right methods, but will use the queue defined outside of this script, if it exists.

File: paf-mvp-frontend/src/utils/queue.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+const log = new Log('PAF', '#3bb8c3');
+
+// API for loading PAF asynchronously and run functions regardless
+// whether it gets loaded before or after their script executes.
+//
+// <script src="url/to/PAF-lib.js" async></script>
+// <script>
+//  var PAF = PAF || {};
+//  PAF.queue = PAF.queue || [];
+//  PAF.queue.push(functionToExecuteOncePrebidLoads);
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:12]: `functionToExecuteOncePrebidLoads` => `functionToExecuteOncePAFLoads`

File: paf-mvp-frontend/src/utils/queue.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+/** Interface for processing every pushed commands as soon as possible.  */
+export interface ICommandProcessor {
+  push(...ops: Command[]): void;
+}
+
+/** Type for handling a duck-typing approach on 'push' function.  */
+export type CommandQueue = Command[] | ICommandProcessor;
+
+/**
+ * @param pusher Commands to process or Processor that is already in place.
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:31]: Typo in parameter name

File: paf-mvp-frontend/src/utils/queue.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    return queue;
+  }
+
+  const processor = new CommandProcessor();
+
+  if (queue && Array.isArray(queue)) {
+    processor.push(...queue);
+  }
+
+  return processor;
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:45]: Why do you return the processor? At first it confused me and I lost some time to understand the whole design, now I don't see what's the point of returning it.

File: paf-mvp-frontend/src/utils/queue.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+//
+// In asynchronous code:
+//  export const queue: CommandQueue = window.PAF.queue || [];
+//  window.PAF.queue = processCommands(queue);
+
+/** An operation executed asynchronously after the PAF-lib is loaded. */
+export type Command = () => void;
+
+/** Interface for processing every pushed commands as soon as possible.  */
+export interface ICommandProcessor {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:23]: Why use an interface in this particular case? Just the `CommandProcessor` class would be enough, as the interface is exported but not used elsewhere.

File: paf-mvp-frontend/src/utils/queue.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  push(...ops: Command[]): void;
+}
+
+/** Type for handling a duck-typing approach on 'push' function.  */
+export type CommandQueue = Command[] | ICommandProcessor;
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:28]: 👍

File: paf-mvp-frontend/src/utils/queue.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+//
+// <script src="url/to/PAF-lib.js" async></script>
+// <script>
+//  var PAF = PAF || {};
+//  PAF.queue = PAF.queue || [];
+//  PAF.queue.push(functionToExecuteOncePrebidLoads);
+// </script>
+//
+// In asynchronous code:
+//  export const queue: CommandQueue = window.PAF.queue || [];
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:16]: See my comment in paf-lib.ts about these 2 lines. To be updated accordingly.

File: paf-mvp-frontend/src/lib/paf-lib.ts [hunk 9]
```diff
@@ -757,3 +758,7 @@ export const getAuditLogByDivId = (divId: DivId): AuditLog | undefined => {
   }
   return getAuditLogByTransaction(prebidTransactionId);
 };
+
+// Set up the queue of asynchronous commands
+export const queue: CommandQueue = window.PAF.queue || [];
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/lib/paf-lib.ts:763]: I think it would make more sense to write it this way:
```typescript
export const queue: CommandQueue = window.PAF.queue || new CommandProcessor();
processCommands(queue);
```
Otherwise, what's the point to have the `processCommands` method accept `CommandProcessor` as param?
The way it is written today, it will always fall into the `if (queue instanceof CommandProcessor) {` case.
I think it's error prone to both **export** `queue` constant, and **re-affect** `window.PAF.queue` after the call to `processCommands`. Since I believe it is useless anyway, I suggest to drop it.

File: paf-mvp-frontend/src/utils/queue.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+//  window.PAF.queue = processCommands(queue);
+
+/** An operation executed asynchronously after the PAF-lib is loaded. */
+export type Command = () => void;
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:20]: More evolutive version:
```typescript
export type Command<T = unknown> = () => T
```

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 1 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts:5]: If for some reason you don't want to use `jest.fn()` (but please justify it 😉), consider using a simple attribute rather than a function:
```
func = () => this.called = true;
```
There is no reason that `func` is actually a function, it always returns the same hardcoded value (which happens to be function, but that's unrelated)

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
+import { TestWatcher } from 'jest';
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:1]: Remove unused imports

File: paf-mvp-frontend/src/utils/queue.ts (hunk 7 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/src/utils/queue.ts:28]: Suggested naming to make it even clearer (it took me some time to understand it as it is today):
```typescript
export type DeferedCommands = Command[];
export class ImmediateCommands {
	// This is just a suggested new name for CommandProcessor
	...
}
export type CommandQueue = DeferedCommands | ImmediateCommands;
```

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    cmd2 = new CommandMocker();
+    cmd3 = new CommandMocker();
+  });
+
+  test('processCommands undefined do not crash', () => {
+    lib.queue = processCommands(undefined);
+  });
+
+  test('processCommands empty array do nothing', () => {
+    lib.queue = processCommands(lib.queue);
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:34]: while I understand the previous test only assesses that no exceptions are thrown, I think this one deserves some assertions.

File: paf-mvp-frontend/src/utils/queue.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+// whether it gets loaded before or after their script executes.
+//
+// <script src="url/to/PAF-lib.js" async></script>
+// <script>
+//  var PAF = PAF || {};
+//  PAF.queue = PAF.queue || [];
+//  PAF.queue.push(functionToExecuteOncePrebidLoads);
+// </script>
+//
+// In asynchronous code:
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:15]: I'm not sure "asynchronous code" is the relevant wording. In this code actually the commands will be ran synchronously, one after the other. "In the library" would be more appropriate, I think.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+describe('queue', () => {
+  let cmd1: CommandMocker;
+  let cmd2: CommandMocker;
+  let cmd3: CommandMocker;
+  const lib = {
+    queue: [] as CommandQueue,
+  };
+
+  beforeEach(() => {
+    cmd1 = new CommandMocker();
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:24]: you must re-initialize the queue in the `beforeEach()` otherwise the assertions are not correct.
As a matter of fact, "nested before processor" actually fails.
But I must admit I don't understand why this test exists.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    lib.queue = processCommands(lib.queue);
+    lib.queue.push(cmd3.func());
+
+    expect(cmd1.called).toBeTruthy();
+    expect(cmd2.called).toBeTruthy();
+    expect(cmd3.called).toBeTruthy();
+  });
+
+  test('nested before processor', () => {
+    lib.queue.push(() => {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:69]: This is a very bizarre test. What's the rationale?

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    // in asynchronous script
+    lib.queue = processCommands(lib.queue);
+
+    expect(cmd1.called).toBeTruthy();
+  });
+
+  test('nested after processor', () => {
+    lib.queue = processCommands(lib.queue);
+
+    lib.queue.push(() => {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:82]: Same comment: this shouldn't be something we expect the library to support.

**Review by OlivierChirouze — CHANGES_REQUESTED** [review]
I think it's a great improvement, but it deserves a couple of fixes both in the implementation and in the tests.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+      this.called = true;
+    };
+  }
+}
+
+describe('queue', () => {
+  let cmd1: CommandMocker;
+  let cmd2: CommandMocker;
+  let cmd3: CommandMocker;
+  const lib = {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:19]: You are testing the `queue` attribute, putting it inside a `lib` object only makes it more verbose.
OPT: use `const queue` instead and remove all the `lib.` in tests.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  });
+
+  test('processCommands empty array do nothing', () => {
+    lib.queue = processCommands(lib.queue);
+  });
+
+  test('processCommands 2 commands', () => {
+    lib.queue = processCommands([cmd1.func(), cmd2.func()]);
+
+    expect(cmd1.called).toBeTruthy();
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:40]: Using `jest.fn()` you can test not only that the method was called, but that it was called **once**, which, I believe, is very important.

**Comment by RomainLofaso** [comment]
> if you add the glue code into the PAF lib as i commented in #168 i think you can avoid the need for a queue in paf-lib

Indeed! To confirm if it is what we need.

Also, I think it can be useful when we will want to add a `setConfig()` function like PrebidJS

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 12 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts:25]: Well, what do you want to assert? 
The array is empty.

File: paf-mvp-frontend/src/utils/queue.ts (hunk 13 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/src/utils/queue.ts]: Indeed. But the "In the library" isn't also a good wording because it is on the page in most cases. I reworded.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 14 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: Good catch!
This actually puzzles me for solving the edge case of the nested calls.

It was to handle such nasty code. Maybe we just can't without an inversion of control which make the code less simple. I don't know if it is worth it.

````
       function fromBidWinToAuditLog(bid) {
            let pafObj = bid.meta.paf;
            if (pafObj === undefined) {
                return;
            }

            // Don't know in witch context fromBidWinToAuditLog
            // will be called. So we push in the queue to be sure
            // that it appears. 
            PAF.queue.push(() => {
                PAF.registerTransmissionResponse({
                    prebidTransactionId: bid.transactionId,
                    adUnitCode: bid.adUnitCode,
                    contentId: pafObj.contentId
                }, pafObj.transmission);
            });
        }

        var pbjs = pbjs || {};
        pbjs.que = pbjs.que || [];

        var PAF = PAF || {};
        PAF.queue = PAF.queue || [];

         pbjs.que.push(function() {
            pbjs.addAdUnits(adUnits);
            pbjs.setConfig(...);

            PAF.queue.push(() => {
                pbjs.onEvent("bidWon", fromBidWinToAuditLog);
            });

            pbjs.requestBids({
                bidsBackHandler: initAdserver,
                timeout: PREBID_TIMEOUT
            });
        });   
````

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 15 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: See my other comment.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 16 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: See my other comment. For my perspective, if you think in term of queue, in can make sense that we add a command/operation inside another command.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 18 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: Indeed. A counter is better.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 1 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: I don't understand your point 🤨.
I was not discussing whether it should be mocked or not: you _are_ already mocking it, just not using the standard way to do it with Jest.

File: paf-mvp-frontend/src/utils/queue.ts (hunk 5 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/src/utils/queue.ts:47]: If you adapt the declaration of queue as I suggest in https://github.com/prebid/paf-mvp-implementation/pull/169#discussion_r877500884, this is not needed (and I insist, I think it is a better pattern to initialize it "the right way" rather than to re-affect it after a call to this method)

File: paf-mvp-frontend/src/lib/paf-lib.ts (hunk 9 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/src/lib/paf-lib.ts:763]: After double thinking, I think you need to update the global variable and make sure it's a processor:

```typescript
if (window.PAF.queue) {
	processCommands(window.PAF.queue);
}
// Reset it once it's been processed
window.PAF.queue = new CommandProcessor();
```

File: paf-mvp-frontend/src/utils/queue.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  const processor = new ImediateCommandProcessor();
+
+  if (queue && Array.isArray(queue)) {
+    processor.push(...queue);
+  }
+
+  return processor;
+};
+
+class ImediateCommandProcessor implements IImediateCommandProcessor {
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:50]: Immediate with two `m` 😛

File: paf-mvp-frontend/src/utils/queue.ts [hunk 20]
```diff
@@ -9,34 +9,36 @@ const log = new Log('PAF', '#3bb8c3');
...
 //
-// In asynchronous code:
+//  Call of the PAF-lib when loaded asynchronously:
 //  export const queue: CommandQueue = window.PAF.queue || [];
 //  window.PAF.queue = processCommands(queue);
 
 /** An operation executed asynchronously after the PAF-lib is loaded. */
 export type Command = () => void;
 
+export type DeferredCommand = Command[];
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/src/utils/queue.ts:22]: This should be a plural

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 14 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: I this case, I guess the processCommands should simply have a `while(queue not empty)` loop?

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 16 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: Yes, thanks for providing an example, now I understand

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 17 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts:9]: I don't see how this is related to nested calls.
See my previous comment: I think what you need is just a `while` loop.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+  beforeEach(() => {
+    cmd1 = jest.fn();
+    cmd2 = jest.fn();
+    cmd3 = jest.fn();
+    lib.queue = [];
+  });
+
+  test('processCommands undefined do not crash', () => {
+    lib.queue = processCommands(undefined);
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:21]: the affectation plays no role here, suggestion:
```typescript
processCommands(undefined);
// Assert no exception thrown
```

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 12 above)
Comment (OlivierChirouze) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts:25]: Then I recommend:
```typescript
processCommands([]);
// Assert no exception thrown
```
(the affectation is useless, and explicit `[]` is more explicit than `lib.queue`

File: paf-mvp-frontend/tests/test-cases/queue.test.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    expect(cmd3.mock.calls.length).toBe(1);
+  });
+
+  test('documented setup', () => {
+    // directly in page
+    lib.queue.push(cmd1);
+    lib.queue.push(cmd2);
+
+    // in asynchronous script
+    lib.queue = processCommands(lib.queue);
```
Comment (OlivierChirouze) [inline paf-mvp-frontend/tests/test-cases/queue.test.ts:51]: This is an illustration of what I dislike about this approach.
This method does two things: it actually processes commands as expected, but also returns a new object that callers must re-affect to `lib.queue`, which is IMHO, confusing.

**Comment by Aerion** [comment]
Should the demo index.hb also be modified to load asynchronously paf-lib?

File: paf-mvp-demo-express/src/views/publisher/index.hbs [hunk 23]
```diff
@@ -90,9 +90,41 @@
...
+
         var pbjs = pbjs || {};
         pbjs.que = pbjs.que || [];
 
+        var PAF = PAF || {};
+        PAF.queue = PAF.queue || [];
+
+         pbjs.que.push(function() {
+            pbjs.addAdUnits(adUnits);
+            pbjs.setConfig(...);
```
[the remark this answers is not in the document]
Comment (OlivierChirouze) [reply paf-mvp-demo-express/src/views/publisher/index.hbs:116]: good catch. I think it's just an example but invalid code.

File: paf-mvp-frontend/src/utils/queue.ts (hunk 5 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/src/utils/queue.ts:47]: Let's close comment #169 before this one.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 14 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: Indeed, thanks!

File: paf-mvp-frontend/src/utils/queue.ts (hunk 19 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/src/utils/queue.ts:53]: This is fair. :-) Fixed.

File: paf-mvp-demo-express/src/views/publisher/index.hbs (hunk 23 above)
[the remark this answers is not in the document]
Comment (RomainLofaso) [reply paf-mvp-demo-express/src/views/publisher/index.hbs]: Fixed.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 17 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: Fixed.

File: paf-mvp-frontend/src/utils/queue.ts (hunk 20 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/src/utils/queue.ts:22]: `DeferredCommands` yes :-)

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 21 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: still for the example of the usage

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 12 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: For demonstration purposes, I prefer to keep it like that.

File: paf-mvp-frontend/tests/test-cases/queue.test.ts (hunk 22 above)
Comment (RomainLofaso) [reply paf-mvp-frontend/tests/test-cases/queue.test.ts]: First, I would like to point out that I don't think that this amount of comments is productive for the project. To be discussed informally (again) but IMO, you should adapt your cursor.

I think it makes sense: a one-liner that you copy-paste from the doc is a no-brainer and removes possibilities of errors.

**Comment by RomainLofaso** [comment]
I am not sure but it seems that I cannot create a PR to the main branch while this one is still existing. Therefore, I am closing it. I hope we will still able to comment.
